src/Controller/ControllerFactory.php

```diff
@@ -38,24 +37,19 @@ class ControllerFactory implements ControllerFactoryInterface
      * @param \Psr\Http\Message\ServerRequestInterface $request The request to build a controller for.
      * @return \Cake\Controller\Controller
      * @throws \Cake\Http\Exception\MissingControllerException
-     * @throws \ReflectionException
      */
     public function create(ServerRequestInterface $request): Controller
     {
         $className = $this->getControllerClass($request);
-        if (!$className) {
+        if ($className === null) {
             $this->missingController($request);
         }
-        /** @var string $className */
-        $reflection = new ReflectionClass($className);
-        if ($reflection->isAbstract() || $reflection->isInterface()) {
```

Comment:
As others mentioned this will cause 500s when the framework used to emit 404, this will generally only impact applications that use fallback routing and abstract/interfaces in their controllers.

Does the usage or reflection have any runtime impact? I know that in early PHP5 Reflection could be a performance problem, but I'm not sure that holds true today.

Reply:
Controller reflection check is IMO never a performance impact, as this is always only a singular PHP invocation per request lifecycle. I think this is in the nano-area of relevance.

The main idea of this change was probably to have less complicated code. And somewhere in the middle we can probably find a good solution here.